Nrfx none run wdt error cases on nrf54h

nordic-segl opened the pull request:
Add WDT instances 010, 011, 131 and 132 that exist in nRF54H20.
Enable execution of tests/drivers/watchdog/wdt_error_cases on nRF54H20.

nordic-segl committed:
drivers: wdt: nrf: Add WDT instances that exist in nrf54h20

Add WDT instances no. 010, 011, 131, 132.

Signed-off-by: Sebastian Głąb <sebastian.glab@nordicsemi.no>

nordic-segl committed:
tests: drivers: watchdog: wdt_error_cases: Run test on nRF54H20

Add overaly files for Application and Radio cores.

Configure test according to watchdog features
available on nRF54H20.

Signed-off-by: Sebastian Głąb <sebastian.glab@nordicsemi.no>

57300 commented on dts/common/nordic/nrf54h20.dtsi:
```diff
@@ -168,6 +168,20 @@
 					     <65 NRF_DEFAULT_IRQ_PRIORITY>;
 			};
+
+			wdt010: watchdog@14000 {
```
Please use unique labels in this file, such as:
```suggestion
			cpuapp_wdt010: watchdog@14000 {
```
Then, you can add localized labels to the following files in `dts/arm/nordic`:

```
/* nrf54h20_cpuapp.dtsi */

wdt010: &cpuapp_wdt010 {};
wdt011: &cpuapp_wdt011 {};

/* nrf54h20_cpurad.dtsi */

wdt010: &cpurad_wdt010 {};
wdt011: &cpurad_wdt011 {};
```

nordic-segl replied:
I applied changes. Check if this is what You meant.

anangl commented on boards/nordic/nrf54h20dk/nrf54h20dk_nrf54h20_cpuapp.yaml:
```diff
@@ -14,3 +14,5 @@ flash: 512
 supported:
   - gpio
   - pwm
+  - spi
```
This change is not related to watchdogs.

nordic-segl replied:
I moved it to a separate commit. This change is needed to run following configuration in Twister:
https://github.com/zephyrproject-rtos/zephyr/pull/71677

anangl commented on dts/common/nordic/nrf54h20.dtsi:
```diff
@@ -168,6 +168,20 @@
+
+			cpuapp_wdt010: watchdog@14000 {
+				compatible = "nordic,nrf-wdt";
+				reg = <0x14000 0x1000>;
```
Add also corresponding entries in `validate_base_addresses.c`.

nordic-segl replied:
This is a black magic for me.
I found following defines in HAL
modules/hal/nordic:
nrfx/mdk/nrf54h20_radiocore.h:    #define NRF_WDT010                            NRF_RADIOCORE_WDT010
nrfx/mdk/nrf54h20_radiocore.h:    #define NRF_WDT010                            NRF_RADIOCORE_WDT010
nrfx/mdk/nrf54h20_application.h:    #define NRF_WDT010                            NRF_APPLICATION_WDT010
nrfx/mdk/nrf54h20_application.h:    #define NRF_WDT010                            NRF_APPLICATION_WDT010

anangl replied:
I think we should use `CHECK_DT_REG(cpuapp_wdt010, NRF_APPLICATION_WDT010);` and so on.

nordic-segl replied:
Done. Please confirm it's all what's needed.

anangl replied:
LGTM now, thanks.

anangl commented on dts/common/nordic/nrf54h20.dtsi:
```diff
@@ -293,6 +307,20 @@
 				interrupts = <61 NRF_DEFAULT_IRQ_PRIORITY>;
 			};
+
+			cpurad_wdt010: watchdog@13000 {
```
Please keep the order by node address.

nordic-segl replied:
Done

anangl commented on tests/drivers/watchdog/wdt_error_cases/boards/nrf54h20dk_nrf54h20_cpurad.overlay:
```diff
@@ -0,0 +1,26 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+&uart135 {
+    pinctrl-0 = < &uart136_default >;
+    pinctrl-1 = < &uart136_sleep >;
+};
```
Why?

nordic-segl replied:
I'm not 100% sure this should be here. I add this because:
- I run Twister on App and Radio using the same UART port (on PC side); So it was easier for me to have both cores printing on the same GPIOs. It's enought to run twister twice, each run on other cpu with different UART port. 
- I used secdom FW with STM traces; it prints logs on GPIOs defined in pincontrol for uart135.

I will remove this to have all tests on Radio core printing on the same UART.

anangl commented on tests/drivers/watchdog/wdt_error_cases/src/main.c:
```diff
@@ -51,6 +51,15 @@
 #define MAX_INSTALLABLE_TIMEOUTS (8)
 #define WDT_WINDOW_MAX_ALLOWED   (0x07CFFFFFU)
 #define DEFAULT_OPTIONS          (WDT_OPT_PAUSE_IN_SLEEP | WDT_OPT_PAUSE_HALTED_BY_DBG)
+#elif defined(CONFIG_SOC_NRF54H20)
```
Why not use `#if defined(CONFIG_SOC_NRF54L15) || defined(CONFIG_SOC_NRF54H20)`? The configuration seems to be the same.

nordic-segl replied:
I thought that nrf54l15 supports WDT_FLAG_RESET_SOC
while nrf54h20 supports WDT_FLAG_RESET_CPU_CORE.
Therefore I started with separate configurations.
Test results show that I was wrong.

I will apply change as suggested.

anangl commented on tests/drivers/watchdog/wdt_error_cases/boards/nrf54h20dk_nrf54h20_cpuapp.overlay:
```diff
@@ -0,0 +1,21 @@
+   status = "okay";
+};
+
+&wdt011 {
+   status = "disabled";
+};
```
Disabling nodes disabled by default is redundant.

nordic-segl replied:
Idea behind this is to have a list of WDT instances that can be easily "switched" ON/OFF.

anangl commented on tests/drivers/watchdog/wdt_error_cases/boards/nrf54h20dk_nrf54h20_cpuapp.overlay:
```diff
@@ -0,0 +1,21 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+&wdt010 {
+   status = "okay";
+};
```
nit:
```suggestion
&wdt010 {
	status = "okay";
};
```

nordic-segl replied:
I changed 4xSPACE to a single TAB.

nordic-segl committed:
boards: nordic: nrf54h20dk: Enable SPI driver tests in Twister

Add SPI to list of supported periperals by
nrf54h20 Application and Radio core.

This is needed to run following tests in Twister
https://github.com/zephyrproject-rtos/zephyr/pull/71677

Signed-off-by: Sebastian Głąb <sebastian.glab@nordicsemi.no>

nordic-segl commented on dts/common/nordic/nrf54h20.dtsi:
```diff
@@ -168,6 +168,20 @@
+				compatible = "nordic,nrf-wdt";
+				reg = <0x15000 0x1000>;
+				status = "disabled";
+				interrupts = <21 NRF_DEFAULT_IRQ_PRIORITY>;
```
_The peripheral ID is the interrupt number (IRQ number)
The peripheral ID is given by bits [22:12] of a peripheral address_
Shall I switch IRQ number to 0x15 or can it stay in decimal?

anangl replied:
The way it is now is consistent with other nodes, so let it stay like this.

nordic-segl replied:
Remaining nodes have IRQ given in decimal form.
